### PR TITLE
(#383) add columnsHitIndex in OapFileFormat

### PR DESCRIPTION
**Pull request by TigerSong**
## What changes were proposed in this pull request?
to analyse which columns hit index,  which returns a map (column->IndexType)

## How was this patch tested?
add unit tests.

### CODE REVIEW

File: src/test/scala/org/apache/spark/sql/test/oap/SharedOapContext.scala [hunk 1]
```diff
@@ -30,4 +34,30 @@ trait SharedOapContext extends SharedSQLContext {
...
+              Some(format)
+            case _ => None
+          }
+        case _ => None
+      }
+    }
+
+    sparkPlan.foreach(node => {
+      if (node.isInstanceOf[FilterExec]) {
+        node.foreach(s => {
```
Comment (xuanyuanking) [inline src/test/scala/org/apache/spark/sql/test/oap/SharedOapContext.scala]: node.children.map

File: src/test/scala/org/apache/spark/sql/test/oap/SharedOapContext.scala [hunk 2]
```diff
@@ -30,4 +34,30 @@ trait SharedOapContext extends SharedSQLContext {
...
   protected implicit def sqlConf: SQLConf = sqlContext.conf
+
+  protected def getColumnsHitIndex(sparkPlan: SparkPlan): Map[String, IndexType] = {
+    val ret = new mutable.HashMap[String, IndexType]()
+    def getOapFileFormat(node: SparkPlan): Option[OapFileFormat] = {
+      node match {
+        case f: FileSourceScanExec =>
+          f.relation.fileFormat match {
+            case format: OapFileFormat =>
+              f.inputRDDs()
```
Comment (xuanyuanking) [inline src/test/scala/org/apache/spark/sql/test/oap/SharedOapContext.scala]: Add comments here to explain why should call inputRDD.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 3]
```diff
@@ -78,6 +78,10 @@ private[sql] class OapFileFormat extends FileFormat
   // TODO inferSchema could be lazy computed
   var inferSchema: Option[StructType] = _
   var meta: Option[DataSourceMeta] = _
+  // { a->BTree(a,b,c), d->BTree(d)->(d), e->Bitmap(e), ... }
```
Comment (xuanyuanking) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: change this comment

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 4]
```diff
@@ -78,6 +78,10 @@ private[sql] class OapFileFormat extends FileFormat
...
+  var columnsHitIndex: Map[String, IndexType] = _
```
Comment (xuanyuanking) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: hitIndexColumns

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 5]
```diff
@@ -78,6 +78,10 @@ private[sql] class OapFileFormat extends FileFormat
...
+
+  def getHitIndexColumns: Map[String, IndexType] = this.columnsHitIndex
```
Comment (xuanyuanking) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: This method may return a empty Map, maybe we should add a flag to check whether `buildReaderWithPartitionValues` has been called or not.

File: src/test/scala/org/apache/spark/sql/test/oap/SharedOapContext.scala [hunk 6]
```diff
@@ -30,4 +34,29 @@ trait SharedOapContext extends SharedSQLContext {
   protected lazy val configuration: Configuration = sparkContext.hadoopConfiguration
 
   protected implicit def sqlConf: SQLConf = sqlContext.conf
+
+  protected def getColumnsHitIndex(sparkPlan: SparkPlan): Map[String, IndexType] = {
```
Comment (adrian-wang) [inline src/test/scala/org/apache/spark/sql/test/oap/SharedOapContext.scala]: The API added here lies in SharedOapContext, but this is only for test, maybe cannot satisfy your production need...

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 7]
```diff
@@ -78,6 +78,17 @@ private[sql] class OapFileFormat extends FileFormat
   // TODO inferSchema could be lazy computed
   var inferSchema: Option[StructType] = _
   var meta: Option[DataSourceMeta] = _
+  // map of columns->IndexType
+  var hitIndexColumns: Map[String, IndexType] = _
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: private?

File: src/test/scala/org/apache/spark/sql/test/oap/SharedOapContext.scala (hunk 6 above)
Comment (TigerSong) [reply src/test/scala/org/apache/spark/sql/test/oap/SharedOapContext.scala]: Yes, this is only for test. For our production(bigSQL), we will write a "bridge" class to get the FileFormat and hitIndexColumns from SparkPlan or something else.  The most important is to add  "hitIndexColumns" in OapFileFormat.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala (hunk 7 above)
Comment (TigerSong) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: fixed in next commit...

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/FilterSuite.scala [hunk 8]
```diff
@@ -903,4 +903,34 @@ class FilterSuite extends QueryTest with SharedOapContext with BeforeAndAfterEac
...
+
+    val df2 = sql("SELECT * FROM parquet_test WHERE a = 1 AND b = 'this is row 1'")
+    checkAnswer(df2, Row(1, "this is row 1") :: Nil)
+    val ret2 = getColumnsHitIndex(df2.queryExecution.sparkPlan)
+    assert(ret2.keySet.size == 2 && ret2.contains("a") && ret2.contains("b"))
+    assert(ret2("a") == ret2("b") && ret2("a").toString
+      == BTreeIndex(BTreeIndexEntry(0)::BTreeIndexEntry(1)::Nil).toString)
+    sql("drop oindex idx1 on parquet_test")
+    sql("drop oindex idx2 on parquet_test")
+  }
```
Comment (LuciferYang) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/FilterSuite.scala]: add similar assertion to other test case ?

Comment (xuanyuanking) [reply src/test/scala/org/apache/spark/sql/execution/datasources/oap/FilterSuite.scala]: Yep, good advice. @TigerSong We should reinforce other UT to ensure index hitting.

**Comment by adrian-wang** [comment]
Thanks for explanation, this change looks good to me. You can file another PR to ensure index hitting in other test cases, thanks!

**Pull request merged**